Source/CarthageKit/Project.swift

```diff
@@ -497,31 +510,24 @@ public final class Project {
 				} else if self.useSubmodules {
 					submodule = Submodule(name: project.relativePath, path: project.relativePath, URL: repositoryURLForProject(project, preferHTTPS: self.preferHTTPS), SHA: revision)
 				}
-
+				
 				return .Success(submodule)
-			}
-			.flatMap(.Merge) { submodule -> SignalProducer<(), CarthageError> in
-				if let submodule = submodule {
-					return addSubmoduleToRepository(self.directoryURL, submodule, GitURL(repositoryURL.path!))
-						.startOnQueue(self.gitOperationQueue)
-				} else {
-					return checkoutRepositoryToDirectory(repositoryURL, workingDirectoryURL, revision: revision)
 				}
-			}
-			.on(started: {
-				self._projectEventsObserver.sendNext(.CheckingOut(project, revision))
-			})
-
-		return commitExistsInRepository(repositoryURL, revision: revision)
-			.promoteErrors(CarthageError.self)
-			.flatMap(.Merge) { exists -> SignalProducer<NSURL, CarthageError> in
-				if exists {
-					return .empty
-				} else {
-					return self.cloneOrFetchDependency(project)
+				.flatMap(.Merge) { submodule -> SignalProducer<(), CarthageError> in
```

Comment:
It looks like this `flatMap` and the following `on` can be moved outside of this `flatMap`.

Then this `flatMap` could be replaced with a simple `map` without using `SignalProducer.attempt` and `Result`.